### ISSUE TITLE
Add targets for .Net 4.6.1 and 4.6.2 to nuget package

### ISSUE THREAD

**Issue by i-sinister** [dotnet/reactive#895]
As of 4.1.5 nuget package for System.Reactive has target for .NetFramework 4.6 so when installed in project targeting 4.6 correct versions of dependent packages are added. However because targets for 4.6.1/4.6.2 are missing when packages is installed in project targeting these frameworks NetStandard is pulled into the project. Because of how NetStandard compatibility for 4.6.x is implemented there are issues happening when library with dependency to NetStandard is added via nuget to the project targeting 4.6.x. Even [msdn documentation for NetStandard](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) says that:
> For .NET Framework projects that need to use such libraries, we recommend that you upgrade the project to target .NET Framework 4.7.2 or higher

One of the problems we are facing is that project using System.Reactive is build against "wrong" version of System.Net.Http similar to [problem described here](https://github.com/dotnet/corefx/issues/25773)

Good explanation can be found at [this Rick Strahls post](https://weblog.west-wind.com/posts/2019/Feb/19/Using-NET-Standard-with-Full-Framework-NET) and suggested workaround is to add target framework.

Steps to reproduce:
* Create console project targeting 4.6.2 for example in VS 2017
* Check that version 4.0.0.0 of System.Net.Http.dll is referenced
* Add reference to System.Reactive 4.1.5
* Check that version of System.Net.Http.dll is changed to 4.2.0.0

**Comment by onovotny** [comment]
Due to the way NuGet works, if you have a `net46` target, a project that targets a higher version of .NET Framework will still use the `net46` version as the specific framework takes prcedence over the `netstandard2.0` version.

Can you verify that you're seeing/using the .NET Framework version? None of those should be using System.Net.Http package at all.

**Comment by i-sinister** [comment]
Nuget does not work the way you described. At least not for me or people who experienced problem in the second link. Third link have detailed explanation of what is going on. And ofcourse I tried to reproduce the issue and my experience is summed up in "Steps to reproduce" - version change for System.Net.Http is an indication of the NetStandard reference is picked up.

Btw nuget version used in my VS installation is 4.9.3.5777 (from the output of the $host command)

**Comment by onovotny** [comment]
Have you looked to see if there are other packages you’re referencing that are net standard 2? It’s not Rx. If theres a repro, I’m happy to take a look.

**Comment by i-sinister** [comment]
It is very easy to reproduce. As I wrote in the issue description:
> Steps to reproduce:
> * Create console project targeting 4.6.2 for example in VS 2017
> * Check that version 4.0.0.0 of System.Net.Http.dll is referenced
> * Add reference to System.Reactive 4.1.5
> * Check that version of System.Net.Http.dll is changed to 4.2.0.0

But if you create console project targeting 4.6 and install latest reactive only 2 nuget packages are added and System.Net.Http version is not changed.

**Comment by onovotny** [comment]
Are you using package.config or package reference? Either way, it sounds like a NuGet issue and not Rx. Rx for .NET pulls in System.ValueTuple 4.5 and Threading Task Extensions 4.5.2. Both of those have `net` dependency nodes and would not pull in netstandard anything.

**Comment by i-sinister** [comment]
@onovotny, did you even try to reproduce issue according to steps in my comment, which you asked couple times? I bet you did not. 

> Either way, it sounds like a NuGet issue and not Rx
Well, nuget works like it is designed. It might be not perfect but it does predictable things. This things cause problems if packages are not crafted "correctly". If you'd care to fix the issue you'd check links - they contain enough about information about what and why nuget does.

This issue is about *nuget package System.Reactive*. If this is the wrong place to one it could you please point me where should this issue be reported?

**Comment by onovotny** [comment]
You can try github.com/nuget/home for NuGet issues. However, Rx.NET itself does not add these dependencies when targeting .NET 4.6+.

NuGet *will always* pick a `net` version over a `netstandard` version when it's in the package, and that's true of both lib and dependency references. If something isn't working right, then NuGet would be a good place to start.

**Comment by ahives** [comment]
@i-sinister I followed your repro steps and can confirm that the System.Net.Http package is updated to 4.2.0.0. I was able to repro using JetBrains Rider 2019.1.1. One of the things you might want to try is binding redirect (https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/redirect-assembly-versions). The caveat is that doing so may cause problems with System.Reactive.

**Comment by onovotny** [comment]
Thanks @ahives. I haven’t had a chance to look closely yet. Can you tell which of tre dependencies is causing this?

**Comment by i-sinister** [comment]
@ahives, I know that binding redirects works. This is how I am dealing with it right now. However this is extremly uncomfortable to work with when you writing library (both transient dependencies and nuget packages) - now all your clients have add binding redirects. 

Apart from this [Microsoft itself recommends to not use NetStandard libraries in 4.6.x framework](https://docs.microsoft.com/en-us/dotnet/standard/net-standard)

So bottom line is that adding System.Reactive nuget package to project targeting 4.6.1/4.6.2 should not add dependency to NetStandard.

**Comment by danielcweber** [comment]
My guess: It's due to the csproj being in legacy format. Migrate to the new core csproj-format and target .net 4.6 from there.

**Comment by danielcweber** [comment]
@onovotny I was also able to reproduce this, however, I found the System.Reactive.dll copied to build output to be correct one, the one targeting .net 4.6, unlike stated at the top. Also, [this discussion](https://github.com/dotnet/corefx/issues/25773) talks about this happening for "certain nuget packages", not Rx specifically. I had a hard time figuring out the definite bug from that discussion though - it definitely seems to be the tooling, and apparently it's a regression, having been fixed at some point in between.

**Comment by i-sinister** [comment]
>  I found the System.Reactive.dll copied to build output to be correct one, the one targeting .net 4.6, unlike stated at the top

@danielcweber, I havent stated that System.Reactive.dll version is wrong nor that the problem is happening when project is targeting 4.6. In fact I wrote the opposite in the third comment:
>But if you create console project targeting 4.6 and install latest reactive only 2 nuget packages are added and System.Net.Http version is not changed.

Problem is that project is build against NetStandard libraries when **System.Reactive package** is added to the project targeting **4.6.1/4.6.2**. 

> Also, this discussion talks about this happening for "certain nuget packages", not Rx specifically

Absolutely correct - there are other packages having the same issue. Probably lots of them. I know at least one - Npgsql version 4.x.

> I had a hard time figuring out the definite bug from that discussion though

Why dont you try to reproduce following steps I provided then? You can enabled detailed build logging and check from the logs that NetStanard is used.

>My guess: It's due to the csproj being in legacy format. Migrate to the new core csproj-format and target .net 4.6 from there.

Apart from this is completely unacceptable to convert large code base as a workaround for a bug in a single nuget package this does not work either. I've tried and project is compiled against NetStandard. It is both expected and not desired, because this is about how nuget resolves dependencies and how system.reactive package is created.

**Comment by i-sinister** [comment]
Made a small research and my guess is that even targeting 4.6.x in System.Reactive package will not solve the problem because there is "certain nuget package" referenced that suffers from the same issue - System.Threading.Tasks.Extensions does not target 4.6.x so NetStandard libraries are pulled when added to the project. Here is related issue https://github.com/dotnet/corefx/issues/32276.

**Comment by onovotny** [comment]
Closing as external
